server/src/WebSocketEvents.js

```diff
@@ -290,20 +290,89 @@ eventResponses.mv_r = (clientSocket) => {
  * @param {*} clientSocket
  * @param {String} data
  */
+
+// TODO: Not sure where to put this
+let entityChatCooldownList = [
+    // { id: 1, scope: "GLOBAL" },
+];
+
 eventResponses.chat = (clientSocket, data) => {
-    if (!data) return;
+    // Can't use ChatState here 'cause it's located outside module
+    const CHAT_SCOPES = {
+        LOCAL: {
+            value: "LOCAL",
+            cooldown: 0,
+        },
+        GLOBAL: {
+            value: "GLOBAL",
+            cooldown: 10000,
+        },
+        TRADE: {
+            value: "TRADE",
+            cooldown: 10000,
+        },
+    };
+
+    if (!data || !data.scope || !data.message) return;
+    if (data.message.length > 255) return;
     if (clientSocket.inGame === false) return;
 
+    const { message, scope } = data;
     const { entity } = clientSocket;
 
     // Ignore this event if they are dead.
     if (entity.hitPoints <= 0) return;
 
+    const { cooldown } = Object
+        .values(CHAT_SCOPES)
+        .find((_scope) => _scope.value === scope);
+
+    // if no cooldown is found it means the client is passing some strange values
+    if (cooldown === undefined) return;
+
+    const date = new Date();
+    const dataToBroadCast = {
+        id: entity.id,
+        displayName: entity.displayName,
+        scope,
+        message,
+        nextAvailableDate: new Date(date.getTime() + cooldown),
+    };
+
+    // send global chats
+    if (data.scope !== CHAT_SCOPES.LOCAL.value) {
+        const cooldownIndex = entityChatCooldownList
```

Comment:
I think this can be simplified.

Since it is keeping an entry for each player, looping this will be O(n). You could keep the time of the last message they sent for each scope as a prop on the clientSocket itself, and just check that for the player the message is coming from, so it would be O(1).

This way all of the loop iterations, filtering and timeouts can be avoided.

You are already doign something like this with the `nextAvailableDate`. You can add this to the socket object and compare it with the current time when they send another message.

Reply:
in `wss.on("connection" ...`

below where there is `clientSocket.inGame = false;`
you can add something like this to the socket object, to track the message times per connection/player.
```js
clientSocket.previousMessageTimes = {
    LOCAL: 0,
    GLOBAL: 0,
    TRADE: 0,
};
 ```

Reply:
This is great 💯. I was thinking of something like this but couldn't figured out how.

Reply:
I also changed the implementation in client.